### PR TITLE
feat: add stream Java API

### PR DESCRIPTION
**Pull request by superhx**
- add StreamClient to support create stream and open stream with SINGLE/MULTIPLE writeMode/readMode.
- add Stream to support RecordBatch append/fetch

TODO: replace old Java API / implement.

### CODE REVIEW

File: sdks/java/lib/src/main/java/sdk/elastic/stream/api/StreamClient.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+     * @return {@link Stream}.
+     */
+    CompletableFuture<Stream> openStream(long streamId, OpenStreamOptions options);
+
+    /**
+     * Get KV client.
+     *
+     * @return {@link KVClient}
+     */
+    KVClient kvClient();
```
Comment (Chillax-0v0) [inline sdks/java/lib/src/main/java/sdk/elastic/stream/api/StreamClient.java:36]: Why `kvClient` in `StreamClient`? Is it better to separate them?

File: sdks/java/lib/src/main/java/sdk/elastic/stream/api/Stream.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    /**
+     * Close the stream.
+     */
+    CompletableFuture<Void> close();
+
+    /**
+     * Destroy stream.
+     */
+    CompletableFuture<Void> destroy();
```
[the remark this answers is not in the document]
Comment (superhx) [reply sdks/java/lib/src/main/java/sdk/elastic/stream/api/Stream.java:39]: - close => is the opposite to open, the stream can open again in future.
- destroy => is the opposite to create, means delete, the stream cannot open anymore, and the stream resource will be recycle.

File: sdks/java/lib/src/main/java/sdk/elastic/stream/api/StreamClient.java (hunk 1 above)
Comment (superhx) [reply sdks/java/lib/src/main/java/sdk/elastic/stream/api/StreamClient.java:36]: The KV ability is provided by Elastic Stream. If separate them, then we need another base ElasticStreamClient to create StreamClient and KVClient?

**Review by Chillax-0v0 — APPROVED** [review]

File: sdks/java/lib/src/main/java/sdk/elastic/stream/api/ClientBuilder.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+    ClientBuilder endpoint(String endpoint);
+
+    ClientBuilder connectionTimeout(Duration duration);
+
+    ClientBuilder channelMaxIdleTime(Duration duration);
+
+    ClientBuilder heartbeatInterval(Duration duration);
+
+    ClientBuilder clientAsyncSemaphoreValue(int semaphore);
```
Comment (mooc9988) [inline sdks/java/lib/src/main/java/sdk/elastic/stream/api/ClientBuilder.java:15]: add setter for rangeCacheSize?

File: sdks/java/lib/src/main/java/sdk/elastic/stream/api/RecordBatch.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package sdk.elastic.stream.api;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+
+/**
+ * Record batch.
+ */
+public interface RecordBatch {
```
Comment (zhouxinyu) [inline sdks/java/lib/src/main/java/sdk/elastic/stream/api/RecordBatch.java:9]: Missing baseOffset and lastOffset?

File: sdks/java/lib/src/main/java/sdk/elastic/stream/api/CreateStreamOptions.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package sdk.elastic.stream.api;
+
+import sdk.elastic.stream.utils.Arguments;
+
+public class CreateStreamOptions {
+    private int replicaCount;
```
Comment (mooc9988) [inline sdks/java/lib/src/main/java/sdk/elastic/stream/api/CreateStreamOptions.java:6]: add retentionPeriodMs?

File: sdks/java/lib/src/main/java/sdk/elastic/stream/api/RecordBatch.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+     * @return record count.
+     */
+    int count();
+
+    /**
+     * Get min timestamp of records.
+     *
+     * @return min timestamp of records.
+     */
+    long baseTimestamp();
```
Comment (mooc9988) [inline sdks/java/lib/src/main/java/sdk/elastic/stream/api/RecordBatch.java:23]: min timestamp or the timestamp of the first record?

Comment (superhx) [reply sdks/java/lib/src/main/java/sdk/elastic/stream/api/RecordBatch.java:23]: min timestamp

File: sdks/java/lib/src/main/java/sdk/elastic/stream/api/CreateStreamOptions.java (hunk 5 above)
Comment (superhx) [reply sdks/java/lib/src/main/java/sdk/elastic/stream/api/CreateStreamOptions.java:6]: stream trim in Kafka situation should controlled by Kafka, so currently there is no need add retentionPeriodMs

File: sdks/java/lib/src/main/java/sdk/elastic/stream/api/ClientBuilder.java (hunk 3 above)
Comment (superhx) [reply sdks/java/lib/src/main/java/sdk/elastic/stream/api/ClientBuilder.java:15]: add rangeCacheSize when client support cache

File: sdks/java/lib/src/main/java/sdk/elastic/stream/api/CreateStreamOptions.java (hunk 5 above)
Comment (mooc9988) [reply sdks/java/lib/src/main/java/sdk/elastic/stream/api/CreateStreamOptions.java:6]: requestTimeout and long-pullingTimeout are also needed.

File: sdks/java/lib/src/main/java/sdk/elastic/stream/api/OpenStreamOptions.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
+package sdk.elastic.stream.api;
+
+import sdk.elastic.stream.utils.Arguments;
+
+public class OpenStreamOptions {
+    private WriteMode writeMode = WriteMode.SINGLE;
+    private ReadMode readMode = ReadMode.MULTIPLE;
```
Comment (mooc9988) [inline sdks/java/lib/src/main/java/sdk/elastic/stream/api/OpenStreamOptions.java:7]: add timetout?

Comment (superhx) [reply sdks/java/lib/src/main/java/sdk/elastic/stream/api/OpenStreamOptions.java:7]: always online stream: there is no append fail

File: sdks/java/lib/src/main/java/sdk/elastic/stream/api/RecordBatch.java (hunk 4 above)
Comment (superhx) [reply sdks/java/lib/src/main/java/sdk/elastic/stream/api/RecordBatch.java:9]: - append RecordBatch,  there is no baseOffset and lastOffset in RecordBatch.
- [new add] fetch RecordBatchWithContext, RecordBatchWithContext extend RecordBatch to add baseOffset and lastOffset.

**Pull request merged**